src/uct/ib/base/ib_md.c

```diff
@@ -898,18 +913,30 @@ static ucs_status_t uct_ib_mkey_pack(uct_md_h uct_md, uct_mem_h uct_memh,
             memh->flags |= UCT_IB_MEM_FLAG_ATOMIC_MR;
             ucs_trace("created atomic key 0x%x for 0x%x", memh->atomic_rkey,
                       memh->lkey);
-        } else if (status != UCS_ERR_UNSUPPORTED) {
+        } else if (status == UCS_ERR_UNSUPPORTED) {
+            /* ignore for atomic MR */
+            status = UCS_OK;
+        } else {
             return status;
         }
     }
+
     if (memh->flags & UCT_IB_MEM_FLAG_ATOMIC_MR) {
         atomic_rkey = memh->atomic_rkey;
     } else {
-        atomic_rkey = UCT_IB_INVALID_RKEY;
+        atomic_rkey = UCT_IB_INVALID_MKEY;
     }
 
-    uct_ib_md_pack_rkey(memh->rkey, atomic_rkey, rkey_buffer);
-    return UCS_OK;
+    if ((flags & UCT_MD_MKEY_PACK_FLAG_INVALIDATE) &&
+        (memh->indirect_rkey == UCT_IB_INVALID_MKEY)) {
```

Comment:
seems we pack the direct rkey when memh->indirect_rkey already exists, and this is not what should be done